If check mode enabled and file missing set changed to true 32676

pilou- opened the pull request:
##### SUMMARY
Modify `AnsibleModule.set_*_if_different` methods: when `check_mode` is enabled and file is missing, returns `True` instead of `False`

Unit tests updated.
Fixes #32676: this allows to avoid exception mentioned in the issue.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/basic.py

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel 5fdb39c4d1) last updated 2017/12/15 18:27:27 (GMT +200)
```

##### ADDITIONAL INFORMATION
Thanks to @mscherer and @Spredzy for the distributed triplet programming session.

pilou- committed:
basic.py: add mock to os.path.exists

pilou- committed:
set_*_if_different: if check_mode enabled & file missing: set changed to True

Fixes #32676
Thanks to mscherer and Spredzy for the distributed triplet programming
session!

mscherer commented on test/units/module_utils/basic/test_set_mode_if_different.py:
```diff
@@ -89,6 +92,7 @@ def test_mode_unchanged_when_already_0660(am, mock_stats, mocker, mode, check_mo
     am.check_mode = check_mode
     mocker.patch('os.lstat', side_effect=[mock_stats['after'], mock_stats['after'], mock_stats['after']])
     m_lchmod = mocker.patch('os.lchmod', return_value=None, create=True)
+    mocker.patch('os.path.exists', return_value=True)
```
For consistency, shouldn't you assert that this is called, like os.lchmod ?

pilou- replied:
Following @Spredzy  advice, in order to be consistent with mock of `os.lstat`, `assert not m_path_exists.called` was removed :)
We kept it only when `return_value` of the `os.path.exists` mock is parametrized.

Adding this assert for every `os.path.exists` mock doesn't hurt, what do you think ?

pilou- commented:
CI Errors seems unrelated to these updates:
- `T=osx/10.11/1`
    ```
    2017-12-16 12:23:35 ERROR: Command "scp -i test/cache/id_rsa -o BatchMode=yes -o ServerAliveCountMax=4 -o ServerAliveInterval=15 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -P 10206 -q -r /tmp/ansible-source-2iEwZC.tgz administrator@parallels-1.testing.ansible.com:/tmp" returned exit status 1.
    ```

- `T=cloud/default/3.6/3`
    ```
    2017-12-16 12:46:57 TASK [azure_rm_virtualmachine_extension : Delete virtual network] **************
    2017-12-16 12:48:44 fatal: [localhost]: FAILED! => {"changed": false, "msg": "Error deleting virtual network testVnet - Azure Error: InternalServerError\nMessage: An error occurred."}
    ```

mscherer commented:
shipit

pilou- commented:
One minor update:
```
diff --git a/lib/ansible/module_utils/basic.py b/lib/ansible/module_utils/basic.py
index 01192d0b8e..56cd1bd171 100644
--- a/lib/ansible/module_utils/basic.py
+++ b/lib/ansible/module_utils/basic.py
@@ -1474,9 +1474,7 @@ class AnsibleModule(object):
         return changed
 
     def check_file_absent_if_check_mode(self, file_path):
-        if not os.path.exists(file_path) and self.check_mode:
-            return True
-        return False
+        return os.path.exists(file_path) and self.check_mode:
 
     def set_directory_attributes_if_different(self, file_args, changed, diff=None, expand=True):
         return self.set_fs_attributes_if_different(file_args, changed, diff, expand)
```

abadger commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1450,6 +1473,9 @@ def set_fs_attributes_if_different(self, file_args, changed, diff=None, expand=T
         return changed
 
+    def check_file_absent_if_check_mode(self, file_path):
+        return not os.path.exists(file_path) and self.check_mode
```
Tiny optimization: check self.check_mode before checking os.path.exists.  That way in the common case we don't have to use a system call to determine if the path exists.

pilou- replied:
Updated, thanks.

abadger commented:
Merged to devel and cherrypicked for 2.4.3rc1
